builder/etcd/etcd.go

```diff
@@ -0,0 +1,524 @@
+		time.Sleep(250 * time.Millisecond)
+	}
+	log.Errf(c, "Etcd is not answering after %d attempts.", count)
+	return false, &cookoo.FatalError{"Could not connect to Etcd."}
+}
+
+// Set sets a value in etcd.
+//
+// Params:
+// 	- key (string): The key
+// 	- value (string): The value
+// 	- ttl (uint64): Time to live
+// 	- client (EtcdGetter): Client, usually an *etcd.Client.
+//
+// Returns:
+// 	- *etcd.Result
+func Set(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	key := p.Get("key", "").(string)
+	value := p.Get("value", "").(string)
+	ttl := p.Get("ttl", uint64(20)).(uint64)
+	client := p.Get("client", nil).(Setter)
+
+	res, err := client.Set(key, value, ttl)
+	if err != nil {
+		log.Infof(c, "Failed to set %s=%s", key, value)
+	}
+
+	return res, err
+}
+
+// FindSSHUser finds an SSH user by public key.
+//
+// Some parts of the system require that we know not only the SSH key, but also
+// the name of the user. That information is stored in etcd.
+//
+// Params:
+// 	- client (EtcdGetter)
+// 	- fingerprint (string): The fingerprint of the SSH key.
+//
+// Returns:
+// - username (string)
+func FindSSHUser(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	client := p.Get("client", nil).(Getter)
+	fingerprint := p.Get("fingerprint", nil).(string)
+
+	res, err := client.Get("/deis/builder/users", false, true)
+	if err != nil {
+		log.Warnf(c, "Error querying etcd: %s", err)
+		return "", err
+	} else if res.Node == nil || !res.Node.Dir {
+		log.Warnf(c, "No users found in etcd.")
+		return "", errors.New("Users not found")
+	}
+	for _, user := range res.Node.Nodes {
+		log.Infof(c, "Checking user %s", user.Key)
+		for _, keyprint := range user.Nodes {
+			if strings.HasSuffix(keyprint.Key, fingerprint) {
+				parts := strings.Split(user.Key, "/")
+				username := parts[len(parts)-1]
+				log.Infof(c, "Found user %s for fingerprint %s", username, fingerprint)
+				return username, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("User not found for fingerprint %s", fingerprint)
+}
+
+// StoreHostKeys stores SSH hostkeys locally.
+//
+// First it tries to fetch them from etcd. If the keys are not present there,
+// it generates new ones and then puts the new onces into etcd.
+//
+// Params:
+// 	- client(EtcdGetterSetter)
+// 	- ciphers([]string): A list of ciphers to generate. Defaults are dsa,
+// 		ecdsa, ed25519 and rsa.
+// 	- basepath (string): Base path in etcd (ETCD_PATH).
+// Returns:
+//
+func StoreHostKeys(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	defaultCiphers := []string{"rsa", "dsa", "ecdsa", "ed25519"}
+	client := p.Get("client", nil).(GetterSetter)
+	ciphers := p.Get("ciphers", defaultCiphers).([]string)
+	basepath := p.Get("basepath", "/deis/builder").(string)
+
+	res, err := client.Get("sshHostKey", false, false)
+	if err != nil || res.Node == nil {
+		log.Infof(c, "Could not get SSH host key from etcd. Generating new ones.")
+		genSSHKeys(c)
```

Comment:
it feels to me that this function must return the err it has and you have to handle it here. Basically what to do if it failed to generate the host SSH keys?